src/connection/host_lookup.c

```diff
@@ -0,0 +1,435 @@
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+      return gethostbyname_r (name, ret, buf, buflen, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+      return gethostbyname_r (name, ret, &ht_data);
+#else
+      return gethostbyname (name);
+#error "HAVE_GETHOSTBYNAME_R"
+#endif
+#endif /* !HAVE_GETHOSTBYNAME_R */
+    }
+
+  hp_memcpy = host_lookup_internal (name, NULL, HOSTNAME_TO_IPADDR);
+  memcpy ((void *) ret, (void *) hp_memcpy, sizeof (struct hostent));
+
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+  (*result) = (struct hostent *) malloc (sizeof (struct hostent));
+  memcpy ((void *) *result, (void *) hp_memcpy, sizeof (struct hostent));
+  return 0;
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return ret;
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return 0;
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif /* HAVE_GETHOSTBYNAME_R */
+}
+
+int
+getnameinfo_uhost (struct sockaddr *addr, socklen_t addrlen, char *host, size_t hostlen, char *serv, size_t servlen,
+		   int flags)
+{
+  struct hostent *hp;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return getnameinfo (addr, addrlen, host, hostlen, NULL, 0, NI_NOFQDN);
+    }
+
+  hp = host_lookup_internal (NULL, addr, IPADDR_TO_HOSTNAME);
+
+  if (hp == NULL)
+    {
+      return EAI_NODATA;
+    }
+
+  strcpy (host, hp->h_name);
+  return 0;
+}
+
+int
+getaddrinfo_uhost (char *node, char *service, struct addrinfo *hints, struct addrinfo **res)
+{
+
+  struct hostent *hp = NULL;
+  struct addrinfo results_out;
+  struct sockaddr_in addr_convert;
+  struct in_addr *in_addr_buf;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return getaddrinfo (node, NULL, hints, res);
+    }
+
+  memset (&results_out, 0, sizeof (results_out));
+
+  in_addr_buf = (struct in_addr *) malloc (sizeof (struct in_addr));
+
+  hp = host_lookup_internal (node, NULL, HOSTNAME_TO_IPADDR);
+
+  if (hp == NULL)
+    {
+      free (in_addr_buf);
+      return EAI_NODATA;
+    }
+
+  /*Constitute struct addrinfo for the out parameter res */
+  *res = (struct addrinfo *) malloc (sizeof (struct addrinfo));
+  results_out.ai_flags = hints->ai_flags;
+
+  results_out.ai_family = hints->ai_family;
+
+  results_out.ai_socktype = hints->ai_socktype;
+
+  results_out.ai_protocol = IPPROTO_TCP;
+
+  memcpy (&in_addr_buf->s_addr, hp->h_addr_list[0], sizeof (in_addr_buf->s_addr));
+  memcpy (&addr_convert.sin_addr, &in_addr_buf->s_addr, sizeof (addr_convert.sin_addr));
+  results_out.ai_addr = (struct sockaddr *) malloc (sizeof (struct sockaddr));
```

Comment:
Does results_out.ai_addr also free memory in freeaddrinfo() ?

Reply:
Yes, of course I will apply it. Thank you.